File: lab/javascript/tecnicas/fila_circular.js
class FilaCircular {
    constructor(tamanho) {
        this.tamanho = tamanho;
        this.fila = new Array(tamanho).fill(null);
        this.front = 0;
        this.rear = 0;
        this.quantidade = 0;
    }

    enfileirar(valor) {
        if (this.quantidade === this.tamanho) {
            console.log("Fila cheia!");
            return;
        }

        this.fila[this.rear] = valor;
        this.rear = (this.rear + 1) % this.tamanho;
        this.quantidade++;
    }

    desenfileirar() {
        if (this.quantidade === 0) {
            console.log("Fila vazia!");
            return null;
        }

        const valor = this.fila[this.front];
        this.fila[this.front] = null;
        this.front = (this.front + 1) % this.tamanho;
        this.quantidade--;
        return valor;
    }

    print() {
        console.log(`Fila: [${this.fila.join(", ")}] | front=${this.front} | rear=${this.rear} | qtd=${this.quantidade}`);
    }
}


const fila = new FilaCircular(3);
fila.enfileirar(10);
fila.enfileirar(20);
fila.enfileirar(30);
fila.print();

console.log("Desenfileirado:", fila.desenfileirar());
fila.enfileirar(40);
fila.print();

File: lab/c/tecnicas/fila_circular.c
#include <stdio.h>

#define TAM 5  // tamanho da fila

typedef struct {
    int fila[TAM];
    int front;
    int rear;
    int quantidade;
} FilaCircular;

void inicializar(FilaCircular *f) {
    f->front = 0;
    f->rear = 0;
    f->quantidade = 0;
}

void enfileirar(FilaCircular *f, int valor) {
    if (f->quantidade == TAM) {
        printf("Fila cheia!\n");
        return;
    }

    f->fila[f->rear] = valor;
    f->rear = (f->rear + 1) % TAM;
    f->quantidade++;
}

int desenfileirar(FilaCircular *f) {
    if (f->quantidade == 0) {
        printf("Fila vazia!\n");
        return -1;
    }

    int valor = f->fila[f->front];
    f->front = (f->front + 1) % TAM;
    f->quantidade--;
    return valor;
}

void imprimir(FilaCircular *f) {
    printf("Fila: ");
    for (int i = 0; i < TAM; i++)
        printf("%d ", f->fila[i]);
    printf("| front=%d | rear=%d | qtd=%d\n", f->front, f->rear, f->quantidade);
}

int main() {
    FilaCircular f;
    inicializar(&f);

    enfileirar(&f, 10);
    enfileirar(&f, 20);
    enfileirar(&f, 30);
    imprimir(&f);

    printf("Desenfileirado: %d\n", desenfileirar(&f));
    enfileirar(&f, 40);
    imprimir(&f);

    return 0;
}
